### PR TITLE
emit an audit event with file transfer statistics when an SFTP session is completed

### PR DESCRIPTION
**Pull request by capnspacehook**
When an SFTP session ends a new event will now be emitted listing all the files that were read/written to and the amount of bytes that were transferred per file. More fine grained info on SFTP requests during a session will be continued to be emitted as well.

This also removes some unused SFTP event codes. They were removed [here](https://github.com/gravitational/teleport/pull/23786) in v13 so it should be completely fine to backport this all the way to v14, since we haven't been using these event codes in awhile.

Closes https://github.com/gravitational/customer-sensitive-requests/issues/336.

changelog: emit an audit event with file transfer information when an SFTP session is completed

### CODE REVIEW

**Comment by zmb3** [comment]
Why not restore the original SFTP write event?

File: tool/teleport/common/sftp.go [hunk 1]
```diff
@@ -89,10 +96,31 @@ type allowedOps struct {
 type sftpHandler struct {
 	logger  *log.Entry
 	allowed *allowedOps
-	events  chan<- *apievents.SFTP
+	files   []*trackedFile
+	events  chan<- protoAuditEvent
+}
+
+type trackedFile struct {
+	*os.File
```
Comment (zmb3) [inline tool/teleport/common/sftp.go:104]: I would not embed this. If someone calls the underlying Read or Write methods won't we fail to update the counters?

File: tool/teleport/common/sftp.go [hunk 2]
```diff
@@ -651,6 +688,25 @@ func onSFTP() error {
 		serveErr = trace.Wrap(serveErr)
 	}
 
+	// Send a summary event last
+	summaryEvent := &apievents.SFTPSummary{
```
Comment (zmb3) [inline tool/teleport/common/sftp.go:692]: What happens if there's an error before the transfer completes? Will we not audit in that case?

File: web/packages/teleport/src/services/audit/types.ts [hunk 3]
```diff
@@ -317,6 +317,7 @@ export const eventCodes = {
   USER_TASK_CREATE: 'UT001I',
   USER_TASK_UPDATE: 'UT002I',
   USER_TASK_DELETE: 'UT003I',
+  SFTP_SUMMARY: 'TS021I',
```
Comment (zmb3) [inline web/packages/teleport/src/services/audit/types.ts:320]: Can you move this to the area where the other SFTP events are declared?

As a side note, I think there are a bunch of unused constants in the code for the SFTP events that we used to emit before the PR landed that removed a bunch of them. We should remove them if we're not going to use them.

File: tool/teleport/common/sftp.go (hunk 1 above)
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:104]: `sftp.Handlers` which we implement only returns `io.ReaderAt`s and `io.WriterAt`s, so `Read` and `Write` can't be called. `sftp.Handlers` does close the reader/writer if it implements `io.Closer`, but doesn't attempt to call anything else. So this should be safe.

File: tool/teleport/common/sftp.go (hunk 2 above)
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:692]: A few lines above I just set the error to `serveErr` and send the summary event no matter what. Then I return `serveErr` so this event should always be emitted save for a panic or something

**Comment by capnspacehook** [comment]
The original SFTP write and other events were removed because they were too noisy, reading/writing to a file typically happens in multiple small requests instead of one large one even for small files. This consolidates all info on an SFTP session into one event. The potential downside is if someone keeps an SFTP connection open for a really long time the amount of data transferred would be emitted for a long time. Given the nature of the request for this event though I don't think that's a big concern though.

File: web/packages/teleport/src/services/audit/types.ts (hunk 3 above)
Comment (capnspacehook) [reply web/packages/teleport/src/services/audit/types.ts:320]: That's fair I'll do that. I debating putting it with the other SFTP event codes or putting it at the bottom next to new events but either works really

File: tool/teleport/common/sftp.go (hunk 1 above)
Comment (capnspacehook) [reply tool/teleport/common/sftp.go]: Good point, I removed the embedding so only the needed methods are exposed.

File: api/proto/teleport/legacy/types/events/events.proto [hunk 4]
```diff
@@ -1892,6 +1892,54 @@ message SFTP {
   string Error = 12 [(gogoproto.jsontag) = "error,omitempty"];
 }
 
+// SFTPSummary is emitted at the end of an SFTP transfer
+message SFTPSummary {
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/events/events.proto:1896]: How can I tell if this was a upload or download? Looks at bytes read vs written?

**Review by zmb3 — APPROVED** [review]

File: api/proto/teleport/legacy/types/events/events.proto (hunk 4 above)
Comment (capnspacehook) [reply api/proto/teleport/legacy/types/events/events.proto:1896]: Yep, technically you can write and read to the same file handle so classifying isn't as straightforward as SCP. But for 99% of cases the program that is issuing SFTP requests will likely only be reading or writing to files, so I can label this if needed

File: lib/srv/regular/sftp.go [hunk 5]
```diff
@@ -189,22 +189,51 @@ func (s *sftpSubsys) Start(ctx context.Context,
...
+			// If the next byte is a NULL byte, this event is a SFTPSummary event
+			var lastEvent bool
+			var event apievents.AuditEvent
+			nextByte, err := r.Peek(1)
 			if err != nil {
-				s.log.WithError(err).Warn("Failed to unmarshal SFTP event.")
-				continue
-			}
+				s.log.WithError(err).Warn("Failed to peek SFTP event file.")
+			} else if nextByte[0] == 0x0 {
```
Comment (r0mant) [inline lib/srv/regular/sftp.go:198]: This feels kinda hacky, do we need to invent this custom unmarshal scheme with a magic null byte? Can't we just unmarshal the event header first and look at the event type like we do with other resources?

Comment (capnspacehook) [reply lib/srv/regular/sftp.go:198]: I'll try that tomorrow, that is probably a better idea in hindsight

**Review by r0mant — APPROVED** [review]

File: tool/teleport/common/sftp.go [hunk 6]
```diff
@@ -89,10 +91,34 @@ type allowedOps struct {
 type sftpHandler struct {
 	logger  *log.Entry
 	allowed *allowedOps
-	events  chan<- *apievents.SFTP
+	mtx     sync.Mutex
```
Comment (zmb3) [inline tool/teleport/common/sftp.go:94]: It's not really clear now which fields the mutex is meant to protect.

General practice is to group the mutex and the fields it protects to help convey when you need to acquire the mutex.

```
type sftpHandler struct {
    logger *log.Entry
    allowed *allowedOps

     // mtx protects the fields that follow
     mtx sync.Mutex
     files []*trackedFile

     events chan<- apievents.AuditEvent
}
```

Comment (capnspacehook) [reply tool/teleport/common/sftp.go:94]: Done

**Comment by webvictim** [comment]
Screenshot of the new audit event

image

**Pull request merged**